[stdlib, SE-0485] Array functions with OutputSpan

glessard opened the pull request:
Adds Array and ContiguousArray initializers and functions deferred from "[SE-0485 OutputSpan](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0485-outputspan.md)".

For both `Array` and `ContiguousArray`:
```
public init<E>(capacity: Int, initializingWith: (inout OutputSpan<Element>) throws(E) -> Void) throws(E)
public func append<E>(addingCapacity: Int, initializingWith: (inout OutputSpan<Element>) throws(E) -> Void) throws(E)
```

Partially addresses: rdar://155660454

glessard committed:
[stdlib] internal array initializer with OutputSpan

glessard committed:
[stdlib] add SE-0485 Array and ContiguousArray inits

glessard committed:
[stdlib] add SE-0485 Array.append with OutputSpan

glessard committed:
[stdlib] add SE-0485 ContiguousArray.append with OutputSpan

glessard committed:
[test] OutputSpan-mediated `init()` and `append()` functions

natecook1000 commented on stdlib/public/core/Array.swift:
```diff
@@ -1599,6 +1595,132 @@ extension Array {
+    }
+  }
+
+  /// Creates an array with the specified capacity, then calls the given
+  /// closure with an OutputSpan covering the array's uninitialized memory.
```
Two changes here (that should be applied to the `append` method and the two symbols in `ContiguousArray`):
- Apple docs style uses "and then", since "then" isn't a conjunction (even though what you've written is widely accepted)
- Type names should always be written as plain English ("output span") or in code voice, and code voice is prohibited in abstracts.

```suggestion
  /// Creates an array with the specified capacity, and then calls the given
  /// closure with an output span covering the array's uninitialized memory.
```

natecook1000 replied:
Looks like we should update the docs for point 1 in the unsafe inits, as well!

glessard commented on stdlib/public/core/Array.swift:
```diff
@@ -1530,34 +1530,29 @@ extension Array {
-      start: firstElementAddress, count: _unsafeUninitializedCapacity)
-    defer {
-      // Update self.count even if initializer throws an error.
-      _precondition(
-        initializedCount <= _unsafeUninitializedCapacity,
-        "Initialized count set to greater than specified capacity."
-      )
```
Yes it is!
```
      _precondition(
        0 <= initializedCount && initializedCount <= capacity,
        "OutputSpan capacity overflow"
      )
```

glessard committed:
[stdlib] fix doc-comment style

glessard committed:
[test] test parsing and diagnostics tests

These tests assumed that the stdlib would never change.
Use local functions of the appropriate shape in order to accommodate changes in the standard library.

glessard committed:
[stdlib] explain why we don’t need a defer block

glessard committed:
[stdlib] tweak availability for SwiftCompatibilitySpan

glessard commented:
@swift-ci please test

glessard commented on stdlib/public/core/Array.swift:
```diff
@@ -1598,7 +1598,111 @@ extension Array {
   }
+}
 
+@available(SwiftCompatibilitySpan 5.0, *)
+@_originallyDefinedIn(module: "Swift;CompatibilitySpan", SwiftCompatibilitySpan 6.2)
```
Unfortunately, the `@_originallyDefinedIn` attribute only works on the extension.

glessard commented on stdlib/public/core/Array.swift:
```diff
@@ -1598,7 +1598,111 @@ extension Array {
+  ///       - span: An `OutputSpan` covering uninitialized memory with
+  ///         space for the specified number of elements.
+  @_alwaysEmitIntoClient
+  public init<E: Error>(
+    capacity: Int,
+    initializingWith initializer: (
+      _ span: inout OutputSpan<Element>
+    ) throws(E) -> Void
+  ) throws(E) {
+    try self.init(
+      _uninitializedCapacity: capacity, initializingWith: initializer
+    )
+  }
+
```
Seems like an artifact; will try to remove the extra one.

glessard commented on test/stdlib/Span/OutputSpanTests.swift:
```diff
@@ -246,3 +247,217 @@ suite.test("InlineArray initialization throws")
+}
+
+suite.test("Array append overflow")
+.skip(.wasiAny(reason: "Trap tests aren't supported on WASI."))
```
StdlibUnittest needs maintenance and upkeep.

glessard committed:
[stdlib] address some repetition

glessard committed:
[stdlib] consolidate defer blocks

glessard committed:
[stdlib] improve explanatory comment

glessard commented on stdlib/public/core/Array.swift:
```diff
@@ -1598,7 +1598,111 @@ extension Array {
+    // no need to finalize in a `defer` block, since throwing will cause
+    // changes to be deinitialized when the output span is deinited.
+    try initializer(&span)
+    self._buffer.mutableCount = unsafe span.finalize(for: buffer)
```
I added some testing to exercise this behaviour, and fixed both the initializer and the append functions to be more consistent.

glessard commented:
@swift-ci please smoke test

glessard committed:
[stdlib] simplifications

DougGregor commented:
Oh, yes please! I've been missing these

glessard commented on stdlib/public/core/Array.swift:
```diff
@@ -1596,12 +1598,106 @@ extension Array {
+    initializingWith initializer: (
+      _ span: inout OutputSpan<Element>
+    ) throws(E) -> Void
+  ) throws(E) {
```
Should we have a return value here? It seems like just returning `Void` is unnecessarily inflexible. With a return value we could take advantage of the additional information when compose this functionality into other things.

glessard commented on stdlib/public/core/ContiguousArray.swift:
```diff
@@ -1115,7 +1115,91 @@ extension ContiguousArray {
+
+@available(SwiftCompatibilitySpan 5.0, *)
+@_originallyDefinedIn(module: "Swift;CompatibilitySpan", SwiftCompatibilitySpan 6.2)
+extension ContiguousArray {
```
As above: the `@_originallyDefinedIn` attribute only works on the extension (perhaps it's actually top level), and it has to be paired with its `@available` declaration.

glessard committed:
[stdlib] preserve items appended until a thrown error

natecook1000 commented on stdlib/public/core/Array.swift:
```diff
@@ -1596,12 +1598,101 @@ extension Array {
+  ///   - capacity: The number of elements to allocate
+  ///     space for in the new array.
+  ///   - initializer: A closure that initializes elements and sets the count
+  ///     of the new array.
```
The `count`-setting is implicit in this version of the initializer.
```suggestion
  ///   - initializer: A closure that initializes elements of the new array.
```

glessard replied:
I'll update the doc-comments in a separate PR, thanks for the notes.

natecook1000 commented on stdlib/public/core/Array.swift:
```diff
@@ -1596,12 +1598,101 @@ extension Array {
+  ///         space for the specified number of additional elements.
+  @_alwaysEmitIntoClient
+  public mutating func append<E: Error>(
+    addingCapacity uninitializedCount: Int,
```
Maybe `freeCapacity` would be accurate and descriptive, and distinguish itself from the unadorned `capacity`?

glessard replied:
Yes, this one was hard to name. We want to establish an amount of capacity over the current `count`, but we don't want to promise that we're necessarily re-allocating. Hopefully this isn't confusing. If we find a better name (or if a better name is needed,) then we can revisit this label.

natecook1000 commented on test/stdlib/Span/OutputSpanTests.swift:
```diff
@@ -275,3 +273,275 @@ suite.test("OutputSpan Sendability")
+  expectEqual(I.count, 6)
+}
+
+suite.test("ContiguousArray append overflow")
```
We should have an underflow (?) check for `a.append(addingCapacity: -1) { ... }` etc.

glessard replied:
I'll add explicit tests for underflow in a separate PR; note that quickly-written ones already trap as expected.

glessard commented on stdlib/public/core/Array.swift:
```diff
@@ -1596,12 +1598,101 @@ extension Array {
+  }
+
+  /// Grows the array to have enough capacity for the specified number of
+  /// elements, then calls the closure with an OutputSpan covering the array's
```
I'll update the doc-comments in a separate PR, thanks for the notes.
